File: LC 1395.cpp
class Solution {
    const static int N = 1e5 + 5;
    int tr1[N], tr2[N];
public:
    void update (int tree[], int x, int v) {
        for (; x < N; x += -x & x) {
            tree[x] += v;
        }
    }

    int query (int tree[], int x) {
        int res = 0;
        for (; x; x -= -x & x) {
            res += tree[x];
        }
        return res;
    }

    int numTeams(vector<int>& rating) {
        int n = rating.size();
        int ans = 0;
        for (int i = 0; i < n; ++i) {
            update(tr2, rating[i], 1);
        }
        for (int i = 0; i < n; ++i) {
            int a = rating[i];
            update(tr2, a, -1);
            ans += query(tr1, a - 1) * (query(tr2, N - 1) - query(tr2, a));
            ans += (query(tr1, N - 1) - query(tr1, a)) * query(tr2, a - 1);
            update(tr1, a, 1);
        }
        return ans;
    }
};

File: 2000/LC 1395.go
import (
    "sort"
)

func numTeams(rating []int) (ans int) {
    n := len(rating)

    update := func(a []int, idx, val int) {
        for i := idx; i <= len(a); i += i & (-i) {
            a[i-1] += val
        }
    }

    query := func(a []int, idx int) (res int) {
        for i := idx; i > 0; i -= i & (-i) {
            res += a[i-1]
        }
        return res
    }

    tmp := make([]int, n)
    copy(tmp, rating)
    sort.Ints(tmp)
    m := map[int]int{}
    for i, x := range tmp {
        m[x] = i + 1
    }

    tr1 := make([]int, n+1)
    tr2 := make([]int, n+1)
    for _, x := range rating {
        update(tr1, m[x], 1)
    }

    for _, x := range rating {
        update(tr1, m[x], -1)
        ans += query(tr2, m[x]-1) * (query(tr1, n) - query(tr1, m[x]))
        ans += query(tr1, m[x]-1) * (query(tr2, n) - query(tr2, m[x]))
        update(tr2, m[x], 1)
    }

    return
}
